File: tests/e2e/runtime/type_overrides/typescript/mysql/runtime.test.ts
/**
 * End-to-end runtime tests for type overrides: TypeScript/MySQL.
 *
 * Each test creates a dedicated MySQL database named test_<uuid> for full
 * isolation. Requires the docker-compose MySQL service on port 13306.
 */
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { join } from 'node:path';
import { readFileSync } from 'node:fs';
import { randomBytes } from 'node:crypto';
import mysql from 'mysql2/promise';

import * as queries from './gen/queries';

const FIXTURES = join(__dirname, '../../../../fixtures/type_overrides/mysql');

const MYSQL_HOST = process.env['MYSQL_HOST'] ?? '127.0.0.1';
const MYSQL_PORT = parseInt(process.env['MYSQL_PORT'] ?? '13306', 10);

const ROOT_CONFIG = {
  host: MYSQL_HOST,
  port: MYSQL_PORT,
  user: 'sqltgen',
  password: 'sqltgen',
  database: 'sqltgen_e2e',
};

const TEST_CONFIG = {
  host: MYSQL_HOST,
  port: MYSQL_PORT,
  user: 'sqltgen',
  password: 'sqltgen',
};

// ─── Setup helpers ────────────────────────────────────────────────────────────

async function makeConn(): Promise<{ conn: mysql.Connection; dbName: string }> {
  const dbName = 'test_' + randomBytes(16).toString('hex');

  const admin = await mysql.createConnection(ROOT_CONFIG);
  await admin.execute(`CREATE DATABASE \`${dbName}\``);
  await admin.execute(`GRANT ALL ON \`${dbName}\`.* TO 'sqltgen'@'%'`);
  await admin.end();

  const conn = await mysql.createConnection({ ...TEST_CONFIG, database: dbName });
  const schemaSql = readFileSync(join(FIXTURES, 'schema.sql'), 'utf8');
  for (const stmt of schemaSql.split(';').map(s => s.trim()).filter(Boolean)) {
    await conn.execute(stmt);
  }
  return { conn, dbName };
}

async function teardown(conn: mysql.Connection, dbName: string): Promise<void> {
  await conn.end();
  const admin = await mysql.createConnection(ROOT_CONFIG);
  await admin.execute(`DROP DATABASE IF EXISTS \`${dbName}\``);
  await admin.end();
}

// ─── :one tests ───────────────────────────────────────────────────────────────

describe(':one queries', () => {
  it('insertEvent and getEvent round-trip', async () => {
    const { conn, dbName } = await makeConn();
    try {
      const payload = { type: 'click', x: 10 };
      const meta = { source: 'web' };
      await queries.insertEvent(conn, 'login',
        payload,
        meta,
        'doc-001',
        new Date('2024-06-01T12:00:00Z'),
        new Date('2024-06-01T14:00:00Z'),
        '2024-06-01',
        new Date('1970-01-01T09:00:00Z'));

      const ev = await queries.getEvent(conn, 1);
      assert.ok(ev);
      assert.equal(ev.name, 'login');
      assert.equal(ev.doc_id, 'doc-001');
      assert.deepEqual(ev.payload, payload);
      assert.deepEqual(ev.meta, meta);
      assert.ok(ev.scheduled_at);
    } finally { await teardown(conn, dbName); }
  });

  it('getEvent returns null for unknown id', async () => {
    const { conn, dbName } = await makeConn();
    try {
      assert.equal(await queries.getEvent(conn, 999), null);
    } finally { await teardown(conn, dbName); }
  });
});

// ─── :many tests ──────────────────────────────────────────────────────────────

describe(':many queries', () => {
  it('listEvents returns all events ordered by id', async () => {
    const { conn, dbName } = await makeConn();
    try {
      const ts = new Date('2024-06-01T12:00:00Z');
      await queries.insertEvent(conn, 'alpha', {}, null, 'doc-1', ts, null, null, null);
      await queries.insertEvent(conn, 'beta',  {}, null, 'doc-2', ts, null, null, null);
      await queries.insertEvent(conn, 'gamma', {}, null, 'doc-3', ts, null, null, null);

      const events = await queries.listEvents(conn);
      assert.equal(events.length, 3);
      assert.equal(events[0].name, 'alpha');
      assert.equal(events[1].name, 'beta');
      assert.equal(events[2].name, 'gamma');
    } finally { await teardown(conn, dbName); }
  });

  it('getEventsByDateRange filters correctly', async () => {
    const { conn, dbName } = await makeConn();
    try {
      await queries.insertEvent(conn, 'early', {}, null, 'doc-1', new Date('2024-01-01T10:00:00Z'), null, null, null);
      await queries.insertEvent(conn, 'mid',   {}, null, 'doc-2', new Date('2024-06-01T12:00:00Z'), null, null, null);
      await queries.insertEvent(conn, 'late',  {}, null, 'doc-3', new Date('2024-12-01T15:00:00Z'), null, null, null);

      const events = await queries.getEventsByDateRange(conn,
        new Date('2024-01-01T00:00:00Z'), new Date('2024-07-01T00:00:00Z'));

      assert.equal(events.length, 2);
      assert.equal(events[0].name, 'early');
      assert.equal(events[1].name, 'mid');
    } finally { await teardown(conn, dbName); }
  });
});

// ─── :exec tests ──────────────────────────────────────────────────────────────

describe(':exec queries', () => {
  it('updatePayload changes payload and meta', async () => {
    const { conn, dbName } = await makeConn();
    try {
      await queries.insertEvent(conn, 'test', { v: 1 }, { source: 'web' }, 'doc-1',
        new Date('2024-06-01T12:00:00Z'), null, null, null);

      const updated = { v: 2, changed: true };
      await queries.updatePayload(conn, updated, null, 1);

      const ev = await queries.getEvent(conn, 1);
      assert.ok(ev);
      assert.deepEqual(ev.payload, updated);
      assert.equal(ev.meta, null);
    } finally { await teardown(conn, dbName); }
  });

  it('updateEventDate updates the date', async () => {
    const { conn, dbName } = await makeConn();
    try {
      await queries.insertEvent(conn, 'dated', {}, null, 'doc-1',
        new Date('2024-06-01T12:00:00Z'), null, '2024-01-01', null);

      await queries.updateEventDate(conn, '2024-12-31', 1);

      const ev = await queries.getEvent(conn, 1);
      assert.ok(ev);
      assert.ok(ev.event_date);
    } finally { await teardown(conn, dbName); }
  });
});

// ─── :execrows tests ──────────────────────────────────────────────────────────

describe(':execrows queries', () => {
  it('insertEventRows returns row count', async () => {
    const { conn, dbName } = await makeConn();
    try {
      const n = await queries.insertEventRows(conn, 'rowtest', {}, null, 'doc-1',
        new Date('2024-06-01T12:00:00Z'), null, null, null);
      assert.equal(n, 1);
    } finally { await teardown(conn, dbName); }
  });
});

// ─── projection tests ─────────────────────────────────────────────────────────

describe('projection queries', () => {
  it('findByDate returns the matching event', async () => {
    const { conn, dbName } = await makeConn();
    try {
      await queries.insertEvent(conn, 'dated', {}, null, 'doc-1',
        new Date('2024-06-01T12:00:00Z'), null, '2024-06-15', null);

      const row = await queries.findByDate(conn, '2024-06-15');
      assert.ok(row);
      assert.equal(row.name, 'dated');
    } finally { await teardown(conn, dbName); }
  });

  it('findByDocId returns the matching event', async () => {
    const { conn, dbName } = await makeConn();
    try {
      await queries.insertEvent(conn, 'doctest', {}, null, 'unique-doc-id',
        new Date('2024-06-01T12:00:00Z'), null, null, null);

      const row = await queries.findByDocId(conn, 'unique-doc-id');
      assert.ok(row);
      assert.equal(row.name, 'doctest');
    } finally { await teardown(conn, dbName); }
  });
});

// ─── count tests ──────────────────────────────────────────────────────────────

describe('count queries', () => {
  it('countEvents counts correctly', async () => {
    const { conn, dbName } = await makeConn();
    try {
      for (let i = 1; i <= 3; i++) {
        await queries.insertEvent(conn, `ev${i}`, {}, null, `doc-${i}`,
          new Date(`2024-06-0${i}T00:00:00Z`), null, null, null);
      }

      const row = await queries.countEvents(conn, new Date('2024-01-01T00:00:00Z'));
      assert.ok(row);
      assert.equal(Number(row.total), 3);
    } finally { await teardown(conn, dbName); }
  });
});

File: tests/e2e/runtime/type_overrides/go/mysql/runtime_test.go
package main

import (
	"context"
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"testing"
	"time"

	gen "e2e-type-overrides-go-mysql/gen"
	_ "github.com/go-sql-driver/mysql"
)

const defaultDSN = "sqltgen:sqltgen@tcp(localhost:13306)/sqltgen_e2e?parseTime=true"

func dsn() string {
	if v := os.Getenv("MYSQL_DSN"); v != "" {
		return v
	}
	return defaultDSN
}

func adminDSN() string {
	d := dsn()
	if idx := strings.LastIndex(d, "/"); idx != -1 {
		rest := d[idx+1:]
		if qIdx := strings.Index(rest, "?"); qIdx != -1 {
			return d[:idx+1] + rest[qIdx:]
		}
		return d[:idx+1]
	}
	return d
}

func randomDBName() string {
	b := make([]byte, 8)
	_, _ = rand.Read(b)
	return "test_" + hex.EncodeToString(b)
}

func setupDB(t *testing.T) (*sql.DB, func()) {
	t.Helper()

	admin, err := sql.Open("mysql", adminDSN())
	if err != nil {
		t.Fatalf("open admin connection: %v", err)
	}

	dbName := randomDBName()
	if _, err := admin.Exec("CREATE DATABASE " + dbName); err != nil {
		admin.Close()
		t.Fatalf("create database %s: %v", dbName, err)
	}

	d := dsn()
	var testDSN string
	if idx := strings.LastIndex(d, "/"); idx != -1 {
		rest := d[idx+1:]
		if qIdx := strings.Index(rest, "?"); qIdx != -1 {
			testDSN = d[:idx+1] + dbName + rest[qIdx:]
		} else {
			testDSN = d[:idx+1] + dbName
		}
	} else {
		testDSN = d + "/" + dbName
	}

	db, err := sql.Open("mysql", testDSN)
	if err != nil {
		admin.Exec("DROP DATABASE IF EXISTS " + dbName)
		admin.Close()
		t.Fatalf("open test database: %v", err)
	}

	ddl, err := os.ReadFile("../../../../fixtures/type_overrides/mysql/schema.sql")
	if err != nil {
		db.Close()
		admin.Exec("DROP DATABASE IF EXISTS " + dbName)
		admin.Close()
		t.Fatalf("read schema: %v", err)
	}

	for _, stmt := range strings.Split(string(ddl), ";") {
		stmt = strings.TrimSpace(stmt)
		if stmt == "" {
			continue
		}
		if _, err := db.Exec(stmt); err != nil {
			db.Close()
			admin.Exec("DROP DATABASE IF EXISTS " + dbName)
			admin.Close()
			t.Fatalf("apply schema statement: %v", err)
		}
	}

	cleanup := func() {
		db.Close()
		admin.Exec("DROP DATABASE IF EXISTS " + dbName)
		admin.Close()
	}
	return db, cleanup
}

func mustJSON(v interface{}) string {
	b, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}
	return string(b)
}

func ts(year, month, day, hour, min, sec int) time.Time {
	return time.Date(year, time.Month(month), day, hour, min, sec, 0, time.UTC)
}

var nilTime sql.NullTime

// ─── :one tests ───────────────────────────────────────────────────────────────

func TestInsertAndGetEvent(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	payload := mustJSON(map[string]interface{}{"type": "click", "x": 10})
	meta := mustJSON(map[string]interface{}{"source": "web"})

	if err := gen.InsertEvent(ctx, db, "login", payload, sql.NullString{String: meta, Valid: true},
		"doc-001", ts(2024, 6, 1, 12, 0, 0), nilTime,
		sql.NullTime{Time: ts(2024, 6, 1, 0, 0, 0), Valid: true},
		nilTime); err != nil {
		t.Fatal(err)
	}

	ev, err := gen.GetEvent(ctx, db, 1)
	if err != nil {
		t.Fatal(err)
	}
	if ev == nil {
		t.Fatal("expected event, got nil")
	}
	if ev.Name != "login" {
		t.Errorf("expected name=login, got %s", ev.Name)
	}
	if ev.DocId != "doc-001" {
		t.Errorf("expected doc_id=doc-001, got %s", ev.DocId)
	}
}

func TestGetEventNotFound(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	ev, err := gen.GetEvent(ctx, db, 999)
	if err != nil {
		t.Fatal(err)
	}
	if ev != nil {
		t.Errorf("expected nil, got %+v", ev)
	}
}

// ─── :many tests ──────────────────────────────────────────────────────────────

func TestListEvents(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	for _, name := range []string{"alpha", "beta", "gamma"} {
		docID := fmt.Sprintf("doc-%s", name)
		if err := gen.InsertEvent(ctx, db, name, "{}", sql.NullString{}, docID, ts(2024, 6, 1, 12, 0, 0), nilTime, nilTime, nilTime); err != nil {
			t.Fatal(err)
		}
	}

	events, err := gen.ListEvents(ctx, db)
	if err != nil {
		t.Fatal(err)
	}
	if len(events) != 3 {
		t.Fatalf("expected 3 events, got %d", len(events))
	}
	if events[0].Name != "alpha" || events[1].Name != "beta" || events[2].Name != "gamma" {
		t.Errorf("unexpected order")
	}
}

func TestGetEventsByDateRange(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	insert := func(name string, created time.Time) {
		t.Helper()
		if err := gen.InsertEvent(ctx, db, name, "{}", sql.NullString{}, "doc-"+name, created, nilTime, nilTime, nilTime); err != nil {
			t.Fatal(err)
		}
	}
	insert("early", ts(2024, 1, 1, 10, 0, 0))
	insert("mid", ts(2024, 6, 1, 12, 0, 0))
	insert("late", ts(2024, 12, 1, 15, 0, 0))

	events, err := gen.GetEventsByDateRange(ctx, db, ts(2024, 1, 1, 0, 0, 0), ts(2024, 7, 1, 0, 0, 0))
	if err != nil {
		t.Fatal(err)
	}
	if len(events) != 2 {
		t.Fatalf("expected 2 events, got %d", len(events))
	}
	if events[0].Name != "early" || events[1].Name != "mid" {
		t.Errorf("unexpected order")
	}
}

// ─── :exec tests ──────────────────────────────────────────────────────────────

func TestUpdatePayload(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	if err := gen.InsertEvent(ctx, db, "test", `{"v":1}`, sql.NullString{}, "doc-1",
		ts(2024, 6, 1, 12, 0, 0), nilTime, nilTime, nilTime); err != nil {
		t.Fatal(err)
	}

	if err := gen.UpdatePayload(ctx, db, `{"v":2}`, sql.NullString{}, 1); err != nil {
		t.Fatal(err)
	}

	ev, err := gen.GetEvent(ctx, db, 1)
	if err != nil {
		t.Fatal(err)
	}
	if ev == nil {
		t.Fatal("expected event")
	}
	if ev.Meta.Valid {
		t.Errorf("expected nil meta")
	}
}

func TestUpdateEventDate(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	if err := gen.InsertEvent(ctx, db, "dated", "{}", sql.NullString{}, "doc-1",
		ts(2024, 6, 1, 12, 0, 0), nilTime, sql.NullTime{Time: ts(2024, 1, 1, 0, 0, 0), Valid: true}, nilTime); err != nil {
		t.Fatal(err)
	}

	if err := gen.UpdateEventDate(ctx, db, sql.NullTime{Time: ts(2024, 12, 31, 0, 0, 0), Valid: true}, 1); err != nil {
		t.Fatal(err)
	}
}

// ─── :execrows tests ──────────────────────────────────────────────────────────

func TestInsertEventRows(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	n, err := gen.InsertEventRows(ctx, db, "rowtest", "{}", sql.NullString{}, "doc-1",
		ts(2024, 6, 1, 12, 0, 0), nilTime, nilTime, nilTime)
	if err != nil {
		t.Fatal(err)
	}
	if n != 1 {
		t.Errorf("expected 1 row, got %d", n)
	}
}

// ─── projection tests ─────────────────────────────────────────────────────────

func TestFindByDate(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	if err := gen.InsertEvent(ctx, db, "dated", "{}", sql.NullString{}, "doc-1",
		ts(2024, 6, 1, 12, 0, 0), nilTime, sql.NullTime{Time: ts(2024, 6, 15, 0, 0, 0), Valid: true}, nilTime); err != nil {
		t.Fatal(err)
	}

	row, err := gen.FindByDate(ctx, db, sql.NullTime{Time: ts(2024, 6, 15, 0, 0, 0), Valid: true})
	if err != nil {
		t.Fatal(err)
	}
	if row == nil {
		t.Fatal("expected row, got nil")
	}
	if row.Name != "dated" {
		t.Errorf("expected dated, got %s", row.Name)
	}
}

func TestFindByDocId(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	if err := gen.InsertEvent(ctx, db, "doctest", "{}", sql.NullString{}, "unique-doc-id",
		ts(2024, 6, 1, 12, 0, 0), nilTime, nilTime, nilTime); err != nil {
		t.Fatal(err)
	}

	row, err := gen.FindByDocId(ctx, db, "unique-doc-id")
	if err != nil {
		t.Fatal(err)
	}
	if row == nil {
		t.Fatal("expected row, got nil")
	}
	if row.Name != "doctest" {
		t.Errorf("expected doctest, got %s", row.Name)
	}
}

// ─── count tests ──────────────────────────────────────────────────────────────

func TestCountEvents(t *testing.T) {
	db, cleanup := setupDB(t)
	defer cleanup()
	ctx := context.Background()

	for i := 1; i <= 3; i++ {
		docID := fmt.Sprintf("doc-%d", i)
		if err := gen.InsertEvent(ctx, db, fmt.Sprintf("ev%d", i), "{}", sql.NullString{}, docID, ts(2024, 6, i, 0, 0, 0), nilTime, nilTime, nilTime); err != nil {
			t.Fatal(err)
		}
	}

	row, err := gen.CountEvents(ctx, db, ts(2024, 1, 1, 0, 0, 0))
	if err != nil {
		t.Fatal(err)
	}
	if row == nil {
		t.Fatal("expected row")
	}
	if row.Total != 3 {
		t.Errorf("expected 3, got %d", row.Total)
	}
}
